src/librustc_mir/transform/generator.rs

```diff
@@ -236,10 +237,28 @@ struct TransformVisitor<'tcx> {
 }
 
 impl TransformVisitor<'tcx> {
-    // Make a GeneratorState rvalue
-    fn make_state(&self, idx: VariantIdx, val: Operand<'tcx>) -> Rvalue<'tcx> {
-        let adt = AggregateKind::Adt(self.state_adt_ref, idx, self.state_substs, None, None);
-        Rvalue::Aggregate(box adt, vec![val])
+    // Make a GeneratorState variant assignment. `core::ops::GeneratorState` only has single
+    // element tuple variants, so we can just write to the downcasted first field and then set the
+    // discriminant to the appropriate variant.
+    fn make_state(
+        &self,
+        idx: VariantIdx,
+        val: Operand<'tcx>,
+        source_info: SourceInfo,
+    ) -> impl Iterator<Item = Statement<'tcx>> {
+        let kind = AggregateKind::Adt(self.state_adt_ref, idx, self.state_substs, None, None);
+        assert_eq!(self.state_adt_ref.variants[idx].fields.len(), 1);
+        let ty = self
+            .tcx
+            .type_of(self.state_adt_ref.variants[idx].fields[0].did)
+            .subst(self.tcx, self.state_substs);
+        expand_aggregate(
```

Comment:
This is sharing code with the Deggaregate pass, isn't it? Nice.

I guess an alternative to this would be to just re-run that pass after other passes that could introduce new aggregates? (Same for ConstProp, which however currently does not even seem to share code.)

Reply:
`ConstProp` was changed to not produce individual assignments by deaggregation, but instead just create an aggregate constant that is assigned in one step. I considered using deaggregation there (which would save us that intern temp alloc thing), but that requires a scheme for inserting more statements, which const prop can't handle right now at all.

Reply:
Rerunning the deaggregator again seemed like overkill (and `expand_aggregate` is already used in other passes, this PR just adds another one). We do something similar for eliminating dead blocks. While we could change all this to using more passes, this seems like a separate discussion for the mir-opt team

Reply:
Fair. So we'll home some tuple-typed consts and hopefully codegen can handle them well even though they are still "aggregated" -- but that's a problem it has to live with anyway.